### PR TITLE
Add AppServicePublish capability to all SDK projects that output to exe

### LINKED ISSUE #4397: Add publish capabilities for SDK based ClassLibraries and ConsoleApps
_From @abpiskunov on December 12, 2018 17:36_

Per our discussion with @nguerrera and @davkean we need to switch to new capabilities schema for publish (see below). To accomplish that one of the workitems is to add

- FolderPublish to SDK-based class libraries and console apps , along with WPF and WinForms (if it is not present yet)
- AppServicePublish to SDK-based class libraries and console apps 

This needs to happen in P2 and earlier rather than later since Publish would need to commit and test changes after all capabilities are added.

Capabilities 
 
•         "Publish": all projects that support any kind of publish should contain this capability
o    All .NET project types
•         "FolderPublish": if project supports publishing to folder
o    Class libraries  (classic and core). Note: we don't own Publish for classic class libraries and classic ConsoleApps, ClickOnce does. There should be separate work item tracking ClickOnce support for our publish framework. However we can add capability now and it would light up when ClickOnce adds apublish provider.
o    Console app (classic, core). 
o    WebSites, WAP, Web core
o    WPF (core and classic)
o    WinForms (classic and core)
•         "AppServicePublish": if project is publishable to any web related publish target: Web, AppService, etc 
o    Console apps (classic and core)
o    WebSites, WAP, Web core
 
Potential pending capabilities:
•         "ClickOnce"
o    Classic console apps
o    WPF (classic)
o    WinForms (classic)
•         "MSIX"
o    Console apps (classic and core)
o    WPF (classic and core)
o    WinForms (classic and core)
 
 
 
Providers 
 
•         FolderProvider: AppliesTo("Publish&FolderPublish")
•         AppServiceProvider: AppliesTo("Publish&AppServicePublish")
•         IISProvider: AppliesTo("Publish&Web")
•         AzureVMProvider: AppliesTo("Publish&Web")
•         AppServiceLinuxProvider: AppliesTo("Publish&DotNetCoreWeb")
•         ContainerRegistryProvider:AppliesTo(Publish&AppServicePublish&(Web|DotnetCoreWeb))
•         ClickOnceProvider:AppliesTo(Publish&ClickOnce)
•         MSIXProvider:AppliesTo(Publish&MSIX)
 
 
Workitems:
 
•	Add FolderPublish to WPF and WinForms (classic and core), Class Libraries and Console Apps (classic and core). Some of them already have it , just make sure all of them have. <- CSharp & SDK
•	Add AppServicePublish to Console apps (classic and core) <- CSharp and SDK
•	Add AppServicePublish to WebSites, WAP, and Core Web <- WAP and WebSdk
 
ClickOnce and MSIX I would leave to their owner to request/add when they are ready.

_Copied from original issue: dotnet/sdk#2747_

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes #4397

I've tested this and it works for C# and VB projects, but for some reason doesn't for F# projects. Hopefully someone knows why, or its just my local environment being weird. I can't see F# opting out of `DefineCommonManagedCapabilities` or anything, but I don't really know where to look either.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets [hunk 1]
```diff
@@ -108,8 +108,11 @@
   
     <!-- Publish capability enables the Publish command for the Project -->
     <ProjectCapability Include="Publish"/>
-  </ItemGroup>
 
+    <!-- All Microsoft.NET.Sdk Exe projects, except WPF and WinForms exes, can be published to app service -->
+    <ProjectCapability Include="AppServicePublish" Condition="'$(OutputType)' == 'Exe'" />
```
Comment (nguerrera) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets]: I think maybe we should put FolderPublish here for consistency even though it duplicates sdk targets. We could then remove from sdk and have justone place to maintain.

Did we get confirmation that output type == exe (but not winexe) is the right trigger for all the cases of app service publish? I believe it is right for all the sdk cases, but I'm not the authority on it.

Comment (davidwengier) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets]: > I think maybe we should put FolderPublish here

I agree, but wasn't sure if duplicates are a problem. Will add.

> Did we get confirmation...

I assumed that you guys had worked out the details for this offline. Apologies if I jumped at fixing it too soon, was aiming for Preview 2.

Comment (davidwengier) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets]: > We could then remove from sdk and have just one place to maintain.

https://github.com/dotnet/sdk/issues/2769

**Comment by Pilchie** [comment]
Tagging @TIHan for the F# question.

**Comment by davidwengier** [comment]
It wasn't a failure, the new capabilities just didn't show up (using `DiagnoseCapabilities`) for F# projects for me. If there is nothing known that F# does differently wrt capabilities or design time targets, I'll see about clearing out my experimental hive etc. and seeing if I can get anywhere.

**Comment by nguerrera** [comment]
No blocking issues from me.

**Comment by davidwengier** [comment]
Looks like the F# issues are just experimental hive deployment issues (thanks for being an excellent rubber duck @TIHan), by patching my VS install correctly the problems go away and new capabilities show up as expected.

**Pull request merged**